packages/core-transactions/src/handlers/transaction.ts

```diff
@@ -110,26 +106,50 @@ export abstract class TransactionHandler implements ITransactionHandler {
         this.revertForRecipient(transaction, walletManager);
     }
 
-    protected applyToSender(transaction: Interfaces.ITransaction, walletManager: State.IWalletManager): void {
+    public applyToSender(transaction: Interfaces.ITransaction, walletManager: State.IWalletManager): void {
         const sender: State.IWallet = walletManager.findByPublicKey(transaction.data.senderPublicKey);
-        sender.balance = sender.balance.minus(transaction.data.amount).minus(transaction.data.fee);
-        sender.incrementNonce();
+        const d: Interfaces.ITransactionData = transaction.data;
+
+        if (Utils.isException(d)) {
+            walletManager.logger.warn(
+                `Transaction forcibly applied as an exception: ${transaction.id}.`
+            );
+        }
+
+        this.canBeApplied(transaction, sender);
+
+        const newBalance: Utils.BigNumber = sender.balance.minus(d.amount).minus(d.fee);
+        if (newBalance.isNegative()) {
+            throw new InsufficientBalanceError();
+        }
+
+        if (d.version > 1) {
+            if (!sender.nonce.plus(1).isEqualTo(d.nonce)) {
+                throw new UnexpectedNonceError(d.nonce, sender.nonce, false);
+            }
+
+            sender.nonce = d.nonce;
+        }
+
+        sender.balance = newBalance;
     }
 
-    protected revertForSender(transaction: Interfaces.ITransaction, walletManager: State.IWalletManager): void {
+    public revertForSender(transaction: Interfaces.ITransaction, walletManager: State.IWalletManager): void {
         const sender: State.IWallet = walletManager.findByPublicKey(transaction.data.senderPublicKey);
-        sender.balance = sender.balance.plus(transaction.data.amount).plus(transaction.data.fee);
-        sender.decrementNonce();
+        const d: Interfaces.ITransactionData = transaction.data;
```

Comment:
Please use better names or just destructure properly then you don't get this kind of variable names.

Reply:
Renamed `d` to `data`.